Inflection test cases around irregular.

dereuromark opened the pull request:
Tests for https://github.com/cakephp/cakephp/issues/14665

It shows that only the CamelCase versions have issues around multi words.
Probably due to the regex used.

Only this one fails for plural:

    $this->assertSame('TestCriteria', Inflector::pluralize('TestCriterion'));

And this one for singular:

    $this->assertSame('TestCriterion', Inflector::singularize('TestCriteria'));

dereuromark committed:
Inflection test cases around irregular.

dereuromark committed:
Inflection test cases around irregular.

dereuromark committed:
Add Inflector detection behind the cache.

dereuromark commented:
I found a way to work also for multi words, by using a inflection detector method and normalizing prior to the regex usage.

dereuromark committed:
Add Inflector detection behind the cache.

markstory commented on src/Utility/Inflector.php:
```diff
@@ -274,10 +274,18 @@ public static function pluralize(string $word): string
+            $normalizedWord = Inflector::underscore($word);
+        }
+
+        if (preg_match('/(.*?(?:\\b|_))(' . static::$_cache['irregular']['pluralize'] . ')$/i', $normalizedWord, $regs)) {
```
Could this pattern `(?:\\b|_)` not be modified to `(?:\\b|_|[A-Z])` instead of incurring additional inflection calls?

dereuromark replied:
I checked, 3 failing tests indicate that this might not be enough.

markstory committed:
Use alternate patterns instead of multiple inflection calls

By generating a pattern for both under_score and CamelCase irregular
words we can handle the new scenarios while also optimizing away a few
string allocations per inflector call. This should net a performance
gain but I've not done any microbenchmarks.

markstory commented:
I found another approach that avoids the additional inflections and also removes some redundant pattern string allocations by leveraging the cache property better.

markstory committed:
Fix formatting.

dereuromark commented on src/Utility/Inflector.php:
```diff
@@ -504,4 +520,25 @@ public static function variable(string $string): string
+     * @param string $word Word
+     * @return string|null
+     */
+    protected static function _detectInflection(string $word): ?string
```
We dont need this method then anymore.
If you want to you can also squash.

markstory committed:
Remove unused method.

markstory commented:
I did some quick benchmarks with PhpBench between these changes and master. The results were not always approximately the same which is good.

markstory commented:
PHPBench test I used.

<details>

```php
<?php
declare(strict_types=1);

/**
 * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
 * @link          http://cakephp.org CakePHP(tm) Project
 * @license       http://www.opensource.org/licenses/mit-license.php MIT License
 */
namespace Cake\Test\Benchmark;

use Cake\Utility\Inflector;

/**
 * @BeforeMethods({"init"})
 * @AfterMethods({"shutdown"})
 */
class DiffBench
{
    public function init()
    {
        $this->words = [
            ['axmen', 'axman'],
            ['men', 'man'],
            ['women', 'woman'],
            ['humans', 'human'],
            ['axmen', 'axman'],
            ['men', 'man'],
            ['women', 'woman'],
            ['humans', 'human'],
            ['categorias', 'categoria'],
            ['houses', 'house'],
            ['powerhouses', 'powerhouse'],
            ['Buses', 'Bus'],
            ['buses', 'bus'],
            ['menus', 'menu'],
            ['news', 'news'],
            ['food_menus', 'food_menu'],
            ['Menus', 'Menu'],
            ['FoodMenus', 'FoodMenu'],
            ['quizzes', 'quiz'],
            ['matrix_rows', 'matrix_row'],
            ['matrices', 'matrix'],
            ['vertices', 'vertex'],
            ['indexes', 'index'],
            ['Aliases', 'Alias'],
            ['Aliases', 'Aliases'],
            ['Media', 'Media'],
            ['NodeMedia', 'NodeMedia'],
            ['alumni', 'alumnus'],
            ['bacilli', 'bacillus'],
            ['cacti', 'cactus'],
            ['foci', 'focus'],
            ['fungi', 'fungus'],
            ['nuclei', 'nucleus'],
            ['octopuses', 'octopus'],
            ['radii', 'radius'],
            ['stimuli', 'stimulus'],
            ['syllabi', 'syllabus'],
            ['termini', 'terminus'],
            ['viruses', 'virus'],
            ['people', 'person'],
            ['people', 'people'],
            ['gloves', 'glove'],
            ['crises', 'crisis'],
            ['taxes', 'tax'],
            ['waves', 'wave'],
            ['bureaus', 'bureau'],
            ['cafes', 'cafe'],
            ['roofs', 'roof'],
            ['foes', 'foe'],
            ['cookies', 'cookie'],
            ['wolves', 'wolf'],
            ['thieves', 'thief'],
            ['potatoes', 'potato'],
            ['heroes', 'hero'],
            ['buffaloes', 'buffalo'],
            ['teeth', 'tooth'],
            ['geese', 'goose'],
            ['feet', 'foot'],
            ['objectives', 'objective'],
            ['briefs', 'brief'],
            ['quotas', 'quota'],
            ['curves', 'curve'],
            ['body_curves', 'body_curve'],
            ['metadata', 'metadata'],
            ['files_metadata', 'files_metadata'],
            ['stadia', 'stadia'],
            ['Addresses', 'Address'],
            ['sieves', 'sieve'],
            ['blue_octopuses', 'blue_octopus'],
            ['chefs', 'chef'],
            ['', ''],
            ['pokemon', 'pokemon'],
        ];
    }

    public function shutdown()
    {
        Inflector::reset();
    }

    /**
     * @Revs(1000)
     * @Iterations(5)
     */
    public function benchPluralize()
    {
        foreach ($this->words as $word) {
            Inflector::pluralize($word[1]);
        }
    }

    /**
     * @Revs(1000)
     * @Iterations(5)
     */
    public function benchSingular()
    {
        foreach ($this->words as $word) {
            Inflector::singularize($word[0]);
        }
    }
}
```

</details>
